src/base/bittorrent/session.cpp

```diff
@@ -5137,7 +5137,15 @@ void Session::handleExternalIPAlert(const lt::external_ip_alert *p)
 
 void Session::handleSessionStatsAlert(const lt::session_stats_alert *p)
 {
-    const qreal interval = lt::total_milliseconds(p->timestamp() - m_statsLastTimestamp) / 1000.;
+    if (m_refreshEnqueued)
```

Comment:
Does it provide any advantages to place it here at top?
I suppose it can be place after zero interval check, therefore there is less job to do, am I correct?

Reply:
> Does it provide any advantages to place it here at top?

IMO, it clearly indicates that it should be scheduled independently from the flow of the rest of this function.

Reply:
I still feel wrong about this.
Currently we have multiple very-close events handled in this function and it seems that by moving reschedule code on top, it will continue to enqueue/produce multiple very-close events in the future.

It also feels wrong to schedule the next update/event before the current one is processed. I don't think the current (slight?) processing delay would have a noticeable effect on calculation correctness.

Reply:
Also I think `Session::enqueueRefresh()` could be improved:
```c++
void Session::enqueueRefresh()
{
    //Q_ASSERT(!m_refreshEnqueued);  // remove this

    QTimer::singleShot(refreshInterval(), this, [this] ()
    {
        m_nativeSession->post_torrent_updates();
        m_nativeSession->post_session_stats();
        enqueueRefresh();  // let this function reschedule itself and remove `enqueueRefresh()` call from other places.
    });

    //m_refreshEnqueued = true;  // remove this variable
}
```

Reply:
> I suppose it can be place after zero interval check, therefore there is less job to do, am I correct?

I not quite understand what did you mean by "there is less job to do".
And if you place it after zero interval check it may not be called at all because of `return` that will be above it.

Reply:
>Currently `enqueueRefresh()` is called from 2 places, maybe that was causing the very-close events?

Really only from one of those places.
IIRC, it is exactly the idea of current implementation:
1. Call libtorrent to post torrent updates and session stats together
2. Schedule it only after last of the two corresponding alerts are handled.

So I doubt it causes the very-close events. Most likely the reason is something else. And its search clearly requires more in-depth research. This PR is intended to fix more trivial errors (such as, for example, to prevent a possible division by zero, without delving into why this zero appears).

Reply:
> I still feel wrong about this.

I could modify `calcRate()` to check for zero interval and returns 0 in such a case. So I could keep structure of `Session::handleSessionStatsAlert()` unchanged.

Reply:
OK, just keep it as is and leave it for another PR.